types/src/transaction/authenticator.rs

```diff
@@ -737,16 +729,382 @@ impl fmt::Display for AuthenticationKey {
     }
 }
 
+#[derive(Clone, Debug, Eq, Hash, PartialEq, Serialize, Deserialize)]
+pub struct MultiKeyAuthenticator {
+    public_keys: MultiKey,
+    signatures: Vec<AnySignature>,
+    signatures_bitmap: aptos_bitvec::BitVec,
+}
+
+impl MultiKeyAuthenticator {
+    pub fn new(public_keys: MultiKey, signatures: Vec<(u8, AnySignature)>) -> Result<Self> {
+        ensure!(
+            public_keys.len() < (u8::MAX as usize),
+            "Too many public keys, {}, in MultiKeyAuthenticator.",
+            public_keys.len(),
+        );
+
+        let mut signatures_bitmap = aptos_bitvec::BitVec::with_num_bits(public_keys.len() as u16);
+        let mut any_signatures = vec![];
+
+        for (idx, signature) in signatures {
+            ensure!(
+                (idx as usize) < public_keys.len(),
+                "Signature index is out of public key range, {} < {}.",
+                idx,
+                public_keys.len(),
+            );
+            ensure!(
+                !signatures_bitmap.is_set(idx as u16),
+                "Duplicate signature index, {}.",
+                idx
+            );
+            signatures_bitmap.set(idx as u16);
+            any_signatures.push(signature);
+        }
+
+        Ok(MultiKeyAuthenticator {
+            public_keys,
+            signatures: any_signatures,
+            signatures_bitmap,
+        })
+    }
+
+    pub fn public_keys(&self) -> &MultiKey {
+        &self.public_keys
+    }
+
+    pub fn signatures(&self) -> Vec<(u8, AnySignature)> {
+        let mut values = vec![];
+        for (idx, signature) in
+            std::iter::zip(self.signatures_bitmap.iter_ones(), self.signatures.iter())
+        {
+            values.push((idx as u8, signature.clone()));
+        }
+        values
+    }
+
+    pub fn verify<T: Serialize + CryptoHash>(&self, message: &T) -> Result<()> {
+        ensure!(
+            self.signatures_bitmap.count_ones() as usize == self.signatures.len(),
+            "Mismatch in number of signatures and the number of bits set in the signatures_bitmap, {} != {}.", self.signatures_bitmap.count_ones(), self.signatures.len(),
+        );
+        ensure!(
+            self.signatures.len() >= self.public_keys.signatures_required() as usize,
+            "Not enough signatures for verification, {} < {}.",
+            self.signatures.len(),
+            self.public_keys.signatures_required(),
+        );
+        for (idx, signature) in
+            std::iter::zip(self.signatures_bitmap.iter_ones(), self.signatures.iter())
+        {
+            signature.verify(&self.public_keys.public_keys[idx], message)?;
```

Comment:
we don't verify that `ones` are within the lens of the public_keys? the [idx] here would panic if it's out of index.

also it doesn't seem we verify the signatures_bitmap has expected number of buckets

Reply:
ugh... missing an ensure statement :sigh:

Reply:
Yep, should have:

```
        ensure!(
            self.signatures_bitmap.last_set_bit().is_some(),
            "Expecting at least one signature in the bitmap"
        );
        
        ensure!(
            self.signatures_bitmap.last_set_bit().unwrap() < self.public_keys.len(),
            "Mismatch in the position of the last signature and the number of PKs, {} >= {}.", 
            self.signatures_bitmap.last_set_bit().unwrap(), self.public_keys.len(),
        );
```

Reply:
@zekun000, why do we care about number of buckets? That's an implementation detail of `BitVec`, no? The checks above should suffice, AFAICT.

Reply:
Fix is [here](https://github.com/aptos-labs/aptos-core/pull/10576).